---
title: "`r params$report_title`"
author: "`r params$report_author`"
date: "`r params$report_time`"
output:
  html_document:
    toc: True
params:
  report_title: "Show transcript boundary coverage stats for a single sample"
  report_author: "Anonymus"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  genome_annotation: "s3://encode-public/2019/06/04/8f6cba12-2ebe-4bec-a15d-53f498979de0/gencode.v29.primary_assembly.annotation_UCSC_names.gtf.gz"
  input_path: "ENCFF465VPI.bed"
---

## Setup

```{r}
library(dplyr)
library(tidyr)
library(RCAS)
library(GenomicRanges)
library(GenomicFeatures)
```

## Parse genome annotation

```{r}
remove_gtf <- FALSE
genome_annotation_gtf <- params$genome_annotation
if (substr(params$genome_annotation, 1, 5) == "s3://") {
  genome_annotation_gtf <- "tmp_gtf.gtf.gz"
  remove_gtf <- TRUE
  system(
    paste(
      "aws s3 cp", params$genome_annotation,
      genome_annotation_gtf, "--no-sign-request", sep=" "
    ),
    intern=FALSE
  )
}
```

```{r}
genome_range_features <- importGtf(genome_annotation_gtf)
txdb_features <- getTxdbFeaturesFromGRanges(genome_range_features)
```

## Show features

```{r}
sample_bed <- importBed(filePath=params$input_path, sampleN=5000)
```

```{r}
summary <- summarizeQueryRegions(queryRegions = sample_bed, txdbFeatures = txdb_features)
```

```{r}
N_features = length(sample_bed)
summary %>%
  data.frame() %>%
  mutate(
    percent = round((count / N_features), 3) * 100
  ) %>%
  tibble::rownames_to_column("feature") %>%
  ggplot(aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 3), label = count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', N_features, ')')) + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))
```

## Feature Boundary Coverage Profile

```{r}
base_gff <- importGtf(genome_annotation_gtf)
gtf_genome <- makeTxDbFromGFF(genome_annotation_gtf)
if (remove_gtf) unlink(genome_annotation_gtf)
transcriptCoords <- transcripts(gtf_genome)
```

```{r}
cvgF <- getFeatureBoundaryCoverage(
  queryRegions = sample_bed,
  featureCoords = transcriptCoords,
  flankSize = 500,
  boundaryType = 'fiveprime',
  sampleN = 1000
)
```


```{r}

```

