---
title: "`r params$report_title`"
author: "`r params$report_author`"
date: "`r params$report_time`"
output:
  html_document:
    toc: True
params:
  report_title: "Retrieve ECODE data files"
  report_author: "Anonymus"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  tmp_folder: "tmp"
  genome_annotation: "s3://encode-public/2019/06/04/8f6cba12-2ebe-4bec-a15d-53f498979de0/gencode.v29.primary_assembly.annotation_UCSC_names.gtf.gz"
  encode_files_manifest_s3: "s3://encode-public/encode_file_manifest.tsv"
  encode_local_manifest: "encode_s3_manifest.tsv"
  experiments_of_interest: ["ENCSR767LLP"]
  downsample_reads: 5000
  chromosome: "chr9"
  genomic_range: [94178791, 94178963]
---

## Setup

```{r}
library(dplyr)
library(tidyr)
library(purrr)
library(ggsci)
library(aws.s3)
library(wiggleplotr)
library(GenomicRanges)
library(GenomicFeatures)
library(RCAS)
```

```{r}
encode_files_manifest_s3 <- params$encode_files_manifest_s3
encode_local_manifest <- file.path(params$tmp_folder, params$encode_local_manifest)

if (!dir.exists(params$tmp_folder)) dir.create(params$tmp_folder, showWarnings=FALSE, recursive=TRUE)
```

## Check S3 repo contents

The public ENCODE bucket does not work well with `aws.s3`, unfortunately, so downloading from cli

```{r}
system(
  paste("aws s3 cp", encode_files_manifest_s3, encode_local_manifest, "--no-sign-request", sep=" "),
  intern=FALSE
)
```

```{r}
manifests <- read.csv(encode_local_manifest, sep="\t")
head(manifests)
```

## Get files associated with experiment

```{r}
samples_df <- manifests %>%
  dplyr::filter(dataset %in% paste0("/experiments/", params$experiments_of_interest, "/")) %>%
  dplyr::filter(file_format == "bigWig") %>%
  dplyr::filter(output_type == "plus strand signal of unique reads") %>%
  dplyr::filter(assembly == "GRCh38") %>%
  dplyr::select(accession, dataset, s3_uri) %>%
  mutate(dataset = gsub("/experiments/", "", dataset)) %>%
  mutate(dataset = gsub("/", "", dataset))

samples_df
```

```{r}
download_bw_files <- function(accession, dataset, s3_uri) {
  local_path <- file.path(params$tmp_folder, paste(accession, "bigWig", sep="."))
  system(paste("aws s3 cp", s3_uri, local_path, "--no-sign-request", sep=" "), intern=FALSE)
  data.frame(accession=accession, dataset=dataset, s3_uri=s3_uri, local_path, scaling_factor=1)
}

local_samples_df <- samples_df %>%
  pmap(download_bw_files) %>%
  bind_rows() %>%
  dplyr::select(
    sample_id=accession, track_id=accession, colour_group=dataset, scaling_factor,
    bigWig=local_path
  )

local_samples_df
```

## Parse genome annotation

```{r}
remove_gtf <- FALSE
genome_annotation_gtf <- params$genome_annotation
if (substr(params$genome_annotation, 1, 5) == "s3://") {
  genome_annotation_gtf <- file.path(params$tmp_folder, "main_annotation.gtf.gz")
  system(
    paste("aws s3 cp", params$genome_annotation, genome_annotation_gtf, "--no-sign-request", sep=" "),
    intern=FALSE
  )
  remove_gtf <- TRUE
}
```

```{r}
genome_range_features <- importGtf(genome_annotation_gtf)
if (remove_gtf) unlink(genome_annotation_gtf)
txdb_features <- getTxdbFeaturesFromGRanges(genome_range_features)
txdb_db <- makeTxDbFromGRanges(genome_range_features)
annotated_GR <- exonsBy(txdb_db, by="gene")
```

## Show ENCODE peaks

```{r}
region_query <- GRanges(
  seqnames=params$chromosome,
  ranges=IRanges(start=params$genomic_range[1], end=params$genomic_range[2]),
  strand="+"
)
```

```{r}
transcript_annot_tab <- genome_range_features %>%
  subsetByOverlaps(region_query) %>%
  data.frame() %>%
  distinct(gene_id, gene_name, strand) %>%
  mutate(transcript_id = gene_id)
```

```{r warning=FALSE, fig.height=9.6, fig.width=4.8}
track_data_tab <- local_samples_df %>%
  mutate(
    sample_id = factor(sample_id, levels=local_samples_df$sample_id),
    track_id = factor(track_id, levels=local_samples_df$track_id)
  )

annotated_GR %>%
  subsetByOverlaps(region_query) %>%
  plotCoverage(
    track_data=track_data_tab, transcript_annotations=transcript_annot_tab, 
    flanking_length=c(500, 500), heights=c(4,1), fill_palette=pal_npg()(9),
  )
```

## Download BAM files to show transcript features

```{r}
samples_df <- manifests %>%
  dplyr::filter(dataset %in% paste0("/experiments/", params$experiments_of_interest, "/")) %>%
  dplyr::filter(file_format == "bam") %>%
  dplyr::filter(output_type == "transcriptome alignments") %>%
  dplyr::filter(analysis_step_version.analysis_step.name == "bulk-rna-seq-alignment-step-v-1") %>%
  dplyr::filter(assembly == "GRCh38") %>%
  dplyr::select(accession, dataset, s3_uri) %>%
  mutate(dataset = gsub("/experiments/", "", dataset)) %>%
  mutate(dataset = gsub("/", "", dataset))

retrieve_bed_files <- function(accession, dataset, s3_uri) {
  local_path <- file.path(params$tmp_folder, paste(accession, "bam", sep="."))
  local_bed_path <- file.path(params$tmp_folder, paste(accession, "bed", sep="."))
  system(paste("aws s3 cp", s3_uri, local_path, "--no-sign-request", sep=" "), intern=FALSE)
  system(paste("bedtools bamtobed -i", local_path, ">", local_bed_path, sep=" "), intern=FALSE)
  data.frame(accession=accession, dataset=dataset, s3_uri=s3_uri, local_bed_path)
}

local_samples_df <- samples_df %>%
  pmap(retrieve_bed_files) %>%
  bind_rows()

local_samples_df
```

```{r}
sample_beds <- local_samples_df$local_bed_path %>%
  map(importBed, sampleN=params$downsample_reads) %>%
  setNames(local_samples_df$accession)
```

```{r}
summary <- sample_beds %>%
  map(summarizeQueryRegions, txdb_features) %>%
  map(data.frame) %>%
  map(tibble::rownames_to_column, "feature") %>%
  imap(~mutate(.x, ACCESSION=.y)) %>%
  bind_rows() %>%
  group_by(ACCESSION) %>%
  mutate(N_features = sum(count)) %>%
  ungroup() %>%
  mutate(
    percent = round((count / N_features), 3)
  )

summary
```

```{r}
summary %>%
  ggplot(aes(x = reorder(feature, -percent), y = percent, fill = ACCESSION)) + 
  geom_col(position = position_dodge2(preserve="single")) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = '', y = paste0("Percent overlap")) + 
  scale_fill_npg() +
  theme_bw(base_size=14) + 
  theme(
    axis.text.x = element_text(angle=30, hjust=1),
    legend.position = "none"
  )
```

## Show transcript boundary coverages

```{r}
cvgList <- sample_beds %>%
  map(calculateCoverageProfileList, txdb_features) %>%
  imap(~mutate(.x, ACCESSION=.y)) %>%
  bind_rows()

head(cvgList)
```

```{r}
cvgList %>%
  dplyr::filter(feature != "cds") %>%
  mutate(
    feature = factor(feature, levels=c(
      "promoters", "transcripts", "threeUTRs", "exons", "introns", "fiveUTRs"
    ))
  ) %>%
  ggplot(aes(x=bins, y=meanCoverage, color=ACCESSION, group=ACCESSION)) +
  geom_line() +
  scale_color_npg() +
  theme_bw(base_size=14) +
  facet_wrap(~ feature, ncol=3)
```