---
title: "`r params$report_title`"
author: "`r params$report_author`"
date: "`r params$report_time`"
output:
  html_document:
    toc: True
params:
  report_title: "Calculate UTR-to-exon ratio for a single sample"
  report_author: "Anonymus"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  genome_annotation: "s3://encode-public/2019/06/04/8f6cba12-2ebe-4bec-a15d-53f498979de0/gencode.v29.primary_assembly.annotation_UCSC_names.gtf.gz"
  input_path: "ENCFF465VPI.bed"
  utre_table_path: "ENCFF465VPI_utre.csv"
  utre_figure_path: "ENCFF465VPI_utre.pdf"
---

## Setup

```{r}
library(dplyr)
library(tidyr)
library(tibble)
library(RCAS)
library(GenomicRanges)
library(GenomicFeatures)
library(ggplot2)
library(ggsci)
```

## Parse genome annotation

```{r}
remove_gtf <- FALSE
genome_annotation_gtf <- params$genome_annotation
if (substr(params$genome_annotation, 1, 5) == "s3://") {
  genome_annotation_gtf <- "tmp_gtf.gtf.gz"
  remove_gtf <- TRUE
  system(
    paste(
      "aws s3 cp", params$genome_annotation,
      genome_annotation_gtf, "--no-sign-request", sep=" "
    ),
    intern=FALSE
  )
}
```

```{r}
genome_range_features <- importGtf(genome_annotation_gtf)
if (remove_gtf) unlink(genome_annotation_gtf)
```

## Check if reads overlap exon or UTR

```{r}
transcript_info <- params$input_path %>%
  importBed(filePath=.) %>%
  queryGff(queryRegions=., gffData=genome_range_features) %>%
  dplyr::distinct(
    gene_id, gene_name, gene_type, transcript_type, type, query_name, query_score,
    width
  )
```

```{r}
trx_type_tab <- transcript_info %>%
  dplyr::distinct(gene_name, query_name, type, .keep_all=TRUE) %>%
  group_by(gene_name, type) %>%
  mutate(
    feture_count = n()
  ) %>%
  ungroup() %>%
  arrange(gene_name)

write.csv(trx_type_tab, params$utre_table_path, row.names=FALSE)
```

```{r}
exon_utr_ratio_tab <- trx_type_tab %>%
  dplyr::filter(type %in% c("exon", "UTR")) %>%
  dplyr::select(type, gene_name) %>%
  mutate(
    trn = 1
  ) %>%
  pivot_wider(names_from=type, values_from=trn, values_fn=length) %>%
  mutate(
    ure = UTR / exon
  ) %>%
  arrange(desc(UTR))

head(exon_utr_ratio_tab)
```

```{r}
p <- exon_utr_ratio_tab %>%
  mutate(
    sum_read = UTR + exon,
    ure = UTR / sum_read
  ) %>%
  ggplot(aes(x=sum_read, y=ure)) + 
  geom_point() + 
  labs(x = "Read count", y = "Percent of reads overlapping UTR") + 
  scale_y_continuous(labels = scales::percent) +
  scale_fill_npg() +
  theme_bw(base_size=14) + 
  theme(
    axis.text.x = element_text(angle=30, hjust=1),
    legend.position = "none"
  )

p
```

```{r}
ggsave(params$utre_figure_path, plot=p)
```
